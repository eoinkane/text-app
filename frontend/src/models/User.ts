import axios, { AxiosResponse } from "axios";

interface UserProps {
  firstName?: string;
  lastName?: string;
  id?: number;
}

type valueof<T> = T[keyof T];

export default class User {
  constructor(private data: UserProps) {}

  get<K extends keyof UserProps>(propName: K): valueof<UserProps> {
    return this.data[propName];
  }

  set(update: UserProps): void {
    Object.assign(this.data, update);
  }
  fetch(): void {
    // axios
    //   .get(`http://localhost:3000/users/${this.get("id")}`)
    //   .then((response: AxiosResponse): void => {
    //     this.set(response.data);
    //   });
  }

  save(): void {
    // const id = this.get("id");
    // if (id) {
    //   axios.put(`http://localhost:3000/users/${id}`, this.data);
    // } else {
    //   axios.post(`http://localhost:3000/users`, this.data);
    // }
  }
}
