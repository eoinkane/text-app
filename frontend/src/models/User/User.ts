import axios, { AxiosResponse, AxiosError } from "axios";

interface UserProps {
  username: string;
  firstName: string;
  lastName: string;
  id: number;
}

type valueof<T> = T[keyof T];

export class CreateError extends Error {}

export type CreatePayload = {
  username: string;
  password: string;
  firstName: string;
  lastName: string;
};

export default class User {
  private data: UserProps | undefined;
  constructor() {}

  get<K extends keyof UserProps>(propName: K): valueof<UserProps> {
    if (!this.data) {
      throw new Error("User is undefined");
    }
    return this.data[propName];
  }

  private set(update: UserProps): void {
    this.data = !this.data ? update : Object.assign(this.data, update);
  }

  init(data: UserProps): void {
    this.set(data);
  }

  static async create(payload: CreatePayload): Promise<User> {
    let data: AxiosResponse | AxiosError;
    try {
      const response = await axios.post(
        "http://127.0.0.1:5000/api/users",
        {
          firstName: payload.firstName,
          lastName: payload.lastName,
        },
        {
          auth: {
            username: payload.username,
            password: payload.password,
          },
        }
      );
      data = response;

      const newUser = new User();

      newUser.init(data.data);
      return newUser;
    } catch (err) {
      console.log(err);
      throw new CreateError("Error creating new User");
    }
  }

  // save(): void {
  //   const id = this.get("id");
  //   if (id) {
  //     axios.put(`http://localhost:3000/users/${id}`, this.data);
  //   } else {
  //     axios.post(`http://localhost:3000/users`, this.data);
  //   }
  // }
}
