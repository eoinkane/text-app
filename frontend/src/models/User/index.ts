import type { CreatePayload } from "./User";
export { default } from "./User";
export { CreateError } from "./User";
export type { CreatePayload };
