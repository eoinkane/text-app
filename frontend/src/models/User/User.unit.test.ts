import axios from "axios";
import User from "../User";

jest.mock("axios");
const mockedAxios = axios as jest.Mocked<typeof axios>;

describe("Loading Context", () => {
  it("New User by User.create", () => {
    const testUserObj = {
      username: "test_user",
      id: 0,
      firstName: "test",
      lastName: "user",
    };
    mockedAxios.post.mockImplementationOnce(() =>
      Promise.resolve({ data: testUserObj })
    );

    const user = User.create({
      username: "test_user",
      password: "test_password",
      firstName: "test",
      lastName: "user",
    });
    expect(user).toMatchSnapshot();

    expect;

    expect(mockedAxios.post).toHaveBeenCalledWith(
      "http://127.0.0.1:5000/api/users",
      {
        firstName: testUserObj.firstName,
        lastName: testUserObj.lastName,
      },
      {
        auth: {
          username: "test_user",
          password: "test_password",
        },
      }
    );
  });

  it("New User by User.create", () => {
    const testUserObj = {
      username: "test_user",
      id: 0,
      firstName: "test",
      lastName: "user",
    };

    const user = new User();
    user.init(testUserObj);

    expect(user).toMatchSnapshot();

    expect(user.get("username")).toBe(testUserObj.username);
    expect(user.get("id")).toBe(testUserObj.id);
    expect(user.get("firstName")).toBe(testUserObj.firstName);
    expect(user.get("lastName")).toBe(testUserObj.lastName);
  });

  it("User get returns expected values", () => {
    const testUserObj = {
      username: "test_user",
      id: 0,
      firstName: "test",
      lastName: "user",
    };

    const user = new User();
    user.init(testUserObj);

    expect(user).toMatchSnapshot();

    expect(user.get("username")).toBe(testUserObj.username);
    expect(user.get("id")).toBe(testUserObj.id);
    expect(user.get("firstName")).toBe(testUserObj.firstName);
    expect(user.get("lastName")).toBe(testUserObj.lastName);
  });

  it("User get returns error when user data is undefined", () => {
    const testUserObj = {
      username: "test_user",
      id: 0,
      firstName: "test",
      lastName: "user",
    };

    global.console = {
      log: jest.fn(), // console.log are ignored in tests

      // Keep native behaviour for other methods, use those to print out things in your own tests, not `console.log`
      error: jest.fn(),
      warn: jest.fn(),
      info: jest.fn(),
      debug: jest.fn(),
      ...console,
    };

    const user = new User();

    expect(user).toMatchSnapshot();

    expect(() => user.get("username")).toThrowError("User is undefined");
  });
});
