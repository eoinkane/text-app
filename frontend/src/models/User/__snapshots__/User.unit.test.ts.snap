// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loading Context New User by User.create 1`] = `Promise {}`;

exports[`Loading Context New User by User.create 2`] = `
User {
  "data": Object {
    "firstName": "test",
    "id": 0,
    "lastName": "user",
    "username": "test_user",
  },
}
`;

exports[`Loading Context User get returns error when user data is undefined 1`] = `
User {
  "data": undefined,
}
`;

exports[`Loading Context User get returns expected values 1`] = `
User {
  "data": Object {
    "firstName": "test",
    "id": 0,
    "lastName": "user",
    "username": "test_user",
  },
}
`;
